### PR TITLE
Edge capsule read lane: #[edge] routes compiled to wasm32-wasip1 with CI-proven byte-identity (#1790)

### PR DESCRIPTION
**Pull request by madmax983**
Implements the first slice of #1790: an opt-in `#[edge]` marker that compiles GET routes into a portable `wasm32-wasip1` "edge capsule", with a wire protocol any CDN shim can implement, a mediated KV read seam, transparent fallthrough to the origin, and a CI-enforced byte-identity conformance suite.

## What's in the box

- **`autumn-edge`** (new crate): the capsule runtime — NDJSON-over-stdio wire protocol (`wire.rs`), a dialogue loop that gates version → method → capability before dispatching into the same axum router the origin mounts (`runtime.rs`), the `EdgeCache` extractor and `EdgeKv` seam (`extract.rs`), a wasmi reference host (`host.rs`, `host` feature), and a reusable two-tier conformance harness (`conformance.rs`, `test-support` feature).
- **`autumn-macros`**: `#[edge]` route marker and `edge_routes[]` collection macro. The route macro emits a native-gated companion so one handler source compiles for both targets, and refuses — with spanned, purpose-written errors — every combination the edge lane cannot honour: non-GET methods, `#[secured]`/`#[authorize]`/`#[step_up]`/`#[throttle]` (both attribute orders, including already-expanded marker consts), `#[intercept(...)]`, `Extension<…>` parameters, `#[static_get]`, `#[ws]`, `#[oauth2_callback]`, and unknown `needs(...)` capabilities.
- **`autumn-web`**: an `edge` feature adding `AppBuilder::with_edge_kv` and a `CacheEdgeKv` adapter so the origin serves edge routes natively with the same KV values the capsule reads.
- **`autumn-cli`**: `autumn build --edge` (single invocation builds origin + capsule, forwarding features and package resolution), `--embed` × edge refusal, and two `autumn doctor` checks (`edge_target`, `edge_routes`) that catch missing targets, guarded/intercepted handlers, and unregistered routes before a build.
- **`examples/edge-greeting`**: the reference app — one `handlers.rs` serving both `src/main.rs` (origin) and `src/bin/edge-capsule.rs` (capsule) — carrying the conformance corpus.
- **CI**: an `edge-conformance` job installs `wasm32-wasip1` and runs the corpus; the suite lives in its own `[[test]]` target so the Docker `--ignored` sweep can never pick it up on a runner without the target. The release-image boot gate vendors `autumn-edge` so path-dependency builds keep working.
- **Docs**: `docs/guide/edge.md` (wire protocol spec included), ADR-0011, CHANGELOG, STABILITY, EXAMPLES, doctor skill.

## Acceptance-criteria traceability

| # | Criterion | Evidence |
|---|---|---|
| 1 | One invocation builds origin + capsule | `autumn build --edge` (autumn-cli/src/build.rs); refusals tested in `autumn-cli/tests/integration/edge.rs` (`build_edge_flag_without_edge_routes_fails_before_compiling`, `build_embed_refuses_edge_routes_before_compiling`, both run with an empty `PATH` to pin that the preflight spawns no cargo); the conformance suite builds the capsule with the same cargo invocation the CLI issues |
| 2 | CI-proven byte-identical bodies + headers | `tier_a_the_capsule_reproduces_the_native_edge_lane_byte_for_byte` and `tier_b_the_origin_agrees_with_the_capsule_and_answers_every_decline` (examples/edge-greeting/tests/conformance.rs, CI `edge-conformance` job) over a 16-case divergence corpus; `origin_and_edge_lanes_agree_byte_for_byte` (autumn/tests/integration/edge_native.rs) runs in the plain CI `test` job on all three OSes |
| 3 | Transparent fallthrough, incl. missing capability | 21 wire-level tests in autumn-edge/tests/runtime_io.rs (every decline reason, sentinel handling, HEAD, non-UTF-8 headers, cookie decline, missing-extension net); `the_corpus_exercises_every_fallthrough_reason` asserts corpus completeness; `an_uninjected_seam_declines_with_the_fallthrough_sentinel` proves the origin-side rejection |
| 4 | Stable-trait platform seam, identical handler source | `EdgeKv` trait + `EdgeCache` extractor; `cache_edge_kv_reads_what_insert_cached_wrote`, `with_edge_kv_registers_the_edge_cache_extension_layer`, `the_injected_seam_is_visible_to_a_native_handler` (edge_native.rs); edge-greeting's single `handlers.rs` feeding both bins is the existence proof; `the_capsule_imports_nothing_that_could_reach_a_filesystem_or_a_socket` pins the sandbox |
| 5 | Actionable build/doctor errors | Five trybuild goldens (`edge_on_post`, `edge_with_secured`, `edge_on_static_get`, `edge_with_intercept`, `edge_with_extension`); macro rejection unit tests in autumn-macros/src/route.rs; `doctor_fails_on_edge_route_with_auth_guard`, `doctor_warns_on_unregistered_edge_route`, `doctor_edge_checks_pass_without_edge_routes` (autumn-cli/tests/integration/edge.rs) |

## Review hardening

A multi-angle pre-open review (correctness, security, API/docs, test/CI integrity, cross-runtime divergence — each finding adversarially verified) plus seven post-open Codex review rounds hardened the slice:

- **Lane identity**: `#[edge]` + `#[intercept(...)]` is a compile error; the origin mount strips the same credential headers the wire strips (proven by the `/whoami` conformance case), so a `HeaderMap`-reading handler observes identical headers on both substrates; non-UTF-8 header values and handler-set `Set-Cookie` fall through instead of diverging; HEAD behaviour pinned at the wire level.
- **Sandbox hardening (wasmi reference host)**: fuel metering (a looping capsule becomes a fallthrough, not a hung host), a 256 MiB linear-memory ceiling, chunked host-side copies of guest-supplied ranges, and cumulative stdout/stderr budgets.
- **Extractor discipline**: `Extension<…>` parameters are refused by the macro (direct and nested), and a runtime net converts axum's missing-extension 500 into a fallthrough — safe even on false positives, since the origin re-runs the identical handler. The class-level fix (a sealed extractor whitelist) is specced in #2244.
- **CLI correctness**: `--features` forwarded to the capsule build; package resolution follows `find_binary` for `--bin` and workspace-root invocations while the selector-free preflight stays cargo-free; the implicit binary fallback never selects the capsule helper.
- Remaining polish items are tracked with analysis in #2244; three review threads stay open deliberately as reasoned deferrals.

## Verification

- Full CI green on the head commit (66 checks): lint (pedantic + nursery, `-D warnings`, panic/determinism gates), workspace tests on all three OSes, edge-conformance (wasm32 capsule vs native, full corpus), docs.rs-posture doc build, SemVer, MSRV, release-image boot gates, fuzz, app-variant conformance matrix.
- Suites: autumn-edge 88, autumn-macros 790, autumn-web edge_native 10, autumn-cli edge/build targeted tests — green.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01SvnoMZoTt222Sdea9oW2mX

---
_Generated by [Claude Code](https://claude.ai/code/session_01SvnoMZoTt222Sdea9oW2mX)_

### CODE REVIEW

File: autumn-cli/src/edge_scan.rs [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+
+/// Walk items, collecting `#[edge]`-marked functions and descending into inline
+/// modules (`mod routes { ... }`). `mod routes;` has no items here — its file is
+/// visited separately by the directory walk.
+fn scan_items(items: &[syn::Item], file: &str, scan: &mut EdgeScan) {
+    for item in items {
+        match item {
+            syn::Item::Fn(item_fn) => {
+                if let Some(found) = edge_fn(&item_fn.attrs, &item_fn.sig, file) {
+                    scan.functions.push(found);
```
[the remark this answers is not in the document]
Comment (madmax983) [reply autumn-cli/src/edge_scan.rs]: Real, but deliberately deferred to #2244 rather than fixed here. Evaluating `cfg` correctly means reproducing cargo's feature resolution (default features, transitive enables via `dep/feature`, target cfgs) inside a token-level source scanner — and a half-right evaluator fails in the worse direction: skipping a cfg'd-**in** route means the capsule silently lacks a route the origin serves, which is the exact silent divergence this lane is built to prevent. Today's behavior errs loud instead: a cfg'd-**out** edge route makes the build demand the WASI target/capsule bin, which is an annoying-but-visible failure the author can route around with `--edge`-less builds or by not feature-gating edge routes in this slice. Tracked in #2244 with this analysis.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: autumn-cli/src/doctor.rs [hunk 2]
```diff
@@ -7119,6 +7119,28 @@ pub fn run(opts: DoctorOptions) {
         check_model_private_columns_impl(&found)
     }));
 
+    // 16. Edge capsule toolchain (issue #1790): a project with `#[edge]` routes
+    //     needs the wasm32-wasip1 std library installed or `autumn build` cannot
+    //     emit the capsule. Both the source scan and the toolchain probe run
+    //     inside the task so they overlap with the other checks.
+    tasks.push(Box::new(|| {
+        let scan = crate::edge_scan::resolve_edge_scan(std::path::Path::new("."));
```
[the remark this answers is not in the document]
Comment (madmax983) [reply autumn-cli/src/doctor.rs]: Deferred to #2244 as part of the doctor-edge-resolution item rather than fixed here: `Path::new(".")`-relative scanning is doctor's house pattern across all its source checks (the autumn-web version probe, list-unsubscribe scan, secret-dotenv scan, and the `src/` glob checks all resolve from cwd the same way), because doctor is deliberately a fast, cargo-free preflight that must work in a project whose dependencies aren't even fetched. Making only the two edge checks spawn `cargo metadata` would diverge from that design and double down on the tree-walk duplication already tracked in #2244 — the right fix is one shared, metadata-aware root resolution for doctor as a whole.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: autumn-cli/src/doctor.rs [hunk 3]
```diff
@@ -7119,6 +7119,28 @@ pub fn run(opts: DoctorOptions) {
...
+        check_edge_target_impl(&scan, installed)
+    }));
+
+    // 17. Edge route wiring (issue #1790): an `#[edge]` handler that also
+    //     carries an auth guard fails the build, an unregistered one is never
+    //     served at the edge, and a missing `src/bin/edge-capsule.rs` leaves
+    //     nothing to compile.
+    tasks.push(Box::new(|| {
+        let scan = crate::edge_scan::resolve_edge_scan(std::path::Path::new("."));
+        let capsule_bin_exists = std::path::Path::new(EDGE_CAPSULE_BIN).exists();
```
[the remark this answers is not in the document]
Comment (madmax983) [reply autumn-cli/src/doctor.rs]: Deferred to #2244 with the sibling doctor finding: checking Cargo's resolved bin targets means either spawning `cargo metadata` (doctor is deliberately cargo-free — it must run before dependencies are fetched) or hand-parsing `[[bin]]`/`autobins` semantics from the manifest, and a half-faithful reimplementation of target discovery would trade today's honest convention check for subtle disagreements with cargo. Both mismatch directions you describe fail loud at `autumn build` (whose resolver *is* metadata-based since this PR's round-2 fix), so doctor errs conservative rather than silent. Tracked in #2244 alongside the workspace-root resolution item, since one shared metadata pass would fix both.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: autumn-macros/src/route.rs [hunk 4]
```diff
@@ -243,9 +290,158 @@ pub fn route_macro(
...
+    // `#[intercept(...)]` attrs were already stripped off `input_fn` by
+    // `parse::extract_interceptors`, so the caller passes the extracted list's
+    // emptiness instead of this fn re-scanning attributes.
+    if has_interceptors {
+        return Some(syn::Error::new(span, EDGE_INTERCEPT_ERROR).to_compile_error());
     }
+    if has_extension_param(input_fn) {
+        return Some(syn::Error::new(span, EDGE_EXTENSION_ERROR).to_compile_error());
+    }
+    None
```
[the remark this answers is not in the document]
Comment (madmax983) [reply autumn-macros/src/route.rs]: Real, and it settles a design question rather than asking for another patch: this is the third bypass of the extension-visibility class (direct `Extension<T>` → type alias → whole-`Request` inspection), and each syntactic refusal provably draws the next one. The class-level answer is the one your first comment named — constrain the accepted extractor set with a positive whitelist (a sealed `EdgeExtract` trait implemented for exactly `Path`/`Query`/`HeaderMap`/`EdgeCache`, replacing the open `Handler<_, EdgeState>` bound) — which is a deliberate API redesign of the `EdgeHandler` seam, not a review-round patch on this PR. Deferred to #2244 with that shape as the recommendation; the interim posture is honest about its limits: the macro refusals + missing-extension net catch the accidental cases, the guide's determinism contract already tells authors a handler must be a function of its request, and a handler that *deliberately* branches on middleware-injected extensions is auth-shaped logic the guide sends to the origin. Fix rounds on this PR are concluded — remaining findings land in #2244 for the maintainer to sequence.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: autumn-cli/src/edge_scan.rs [hunk 5]
```diff
@@ -0,0 +1,538 @@
...
+fn registered_idents(stream: &TokenStream) -> Vec<String> {
+    let mut out = Vec::new();
+    let mut current: Option<String> = None;
+    for tree in stream.clone() {
+        match tree {
+            TokenTree::Punct(p) if p.as_char() == ',' => {
+                out.extend(current.take());
+            }
+            TokenTree::Ident(ident) => current = Some(ident.to_string()),
+            _ => {}
```
[the remark this answers is not in the document]
Comment (madmax983) [reply autumn-cli/src/edge_scan.rs]: Real — the bare-identifier match is one of the scanner's documented recognition limits, and same-named edge handlers in two modules would indeed cross-satisfy a registration. Tracked in #2244 (fix rounds on this PR are concluded); the natural fix is matching on the qualified path when the registration provides one, falling back to the bare identifier only for unqualified entries.

---
_Generated by [Claude Code](https://claude.ai/code)_
